### ISSUE TITLE
Flatpak: Blackmagic Libs no longer supported

### ISSUE THREAD

**Issue by JamesMilneFilmlight** [obsproject/obs-studio#6004]
### Operating System Info

Other

### Other OS

Rocky Linux 8.5

### OBS Studio Version

27.2.1

### OBS Studio Log URL

https://obsproject.com/logs/XwTtVbDrBB1juHxu

### Expected Behavior

In the previous release of OBS via flathub (27.1.3) there was support for a com.obsproject.Studio.BlackMagicLibs extension, to provide the Blackmagic DeckLink API library to OBS.

The process for building the extension was documented here:
https://github.com/flathub/com.obsproject.Studio/wiki/BlackMagic-Support

This repository is no longer functional.

The new Flatpak configuration file does not provide support for the BlackMagicLibs extension, and there doesn't appear to be an easy way to expose the libDeckLinkAPI.so to OBS otherwise.

### Current Behavior

OBS 27.2.1 under Flatpak cannot access Blackmagic SDI hardware.

### Steps to Reproduce

1. Install OBS 27.2.1 from flathab
2. Follow instructions at https://github.com/flathub/com.obsproject.Studio/wiki/BlackMagic-Support
3. DeckLink devices are not accessible when you run OBS.

**Comment by tytan652** [comment]
Flatpak support became official since 27.2. I think the license that those libraries are under prevent OBS Project to re-add this method as it was.

And this section of the flathub wiki is for the unofficial Flatpak.

Maybe ask BlackMagic if they can create a official Flatpak. If there is the OBS Studio Flatpak could maybe use it.

**Comment by kkartaltepe** [comment]
@GeorgesStavracas is this something we can support?

**Comment by GeorgesStavracas** [comment]
Hi. It would be possible to integrate those libraries, yes, but unfortunately BlackMagic doesn't allow arbitrarily downloading their files - all their content is blocked behind a login wall, and there's no stable URL to download it from. I've tried to contact them about a month ago about that, but they said nothing could be done about it.

I honestly don't know how to feasibly integrate these libraries, here are some terrible ideas:

 * Spawn a browser window when installing BlackMagic libs, and let users download the file manually (this is a terrible solution UX wise, and won't work on headless setups)
 * Manually download the download page from BlackMagic, and magically figure out the URL (this is super, super fragile and probably won't work due to JavaScript)
 * Automate the Flatpak manifest linked above

I would love to help and support BlackMagic to create official Flatpaks for their libraries, but I personally don't expect that to happen anytime soon.

**Comment by vchernin** [comment]
FTR, people do seem to have written [hacks to download from blackmagic](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=decklink-sdk) in the AUR (I think these are the same libraries?) 

Likely fragile indeed, but could that be considered functional enough as a workaround to make a Flatpak extension for this? I assume these libraries need to be extra-data since they are proprietary/not redistributable? If that’s the case, then it’s probably too broken since flatpak-external-data-checker doesn’t support arbitrary scripts (yet).

On the other hand, a webview is a hack, but I wonder how many other non redistributable pieces of software require a user interactive download page. Because perhaps it would be useful if Flatpak extra-data natively could open a webpage and grab the file after the user clicks a download button.

**Comment by kkartaltepe** [comment]
Im more interested in updating the instructions already provided for the old flatpak, that could work on our flatpak. I know we probably wont be able to distribute Blackmagic ourselves without potentially lots of work technical or otherwise.

**Comment by vchernin** [comment]
Well, if you just wanted to update [the instructions](https://github.com/flathub/com.obsproject.Studio/wiki/BlackMagic-Support) without changing the overall UX:

1. Change the git URL to this repo instead of Flathub.
2. Add [the manifest for the blackmagic extension](https://github.com/flathub/com.obsproject.Studio/blob/master/com.obsproject.Studio.BlackMagicLibs.json) to this repo.
3. Put the instructions themselves somewhere.

This should work provided the OBS manifest here re-added the Blackmagic extension:
```
"com.obsproject.Studio.BlackMagicLibs": {
            "directory": "lib/blackmagic",
            "add-ld-path": ".",
            "no-autodownload": true,
            "autodelete": true
        },
```
[(and probably this line too)](https://github.com/flathub/com.obsproject.Studio/blob/ef97b9e6f0f70060c5cae317664bdb46c2d088bf/com.obsproject.Studio.json#L309)

As before, the user is still manually downloading the library.

Nonetheless, I am interested to see if there's a more user-friendly approach possible rather than telling users to `git clone`.

**Comment by trwnh** [comment]
I just went back to the Arch community build of obs-studio instead of the Flatpak, so I have Decklink support again but I am now missing browser-source and other features that are available in Flatpak.

**Comment by RiQuY** [comment]
Any news on this issue?

I have an Intensity Pro 4K, I followed the steps listed in https://github.com/flathub/com.obsproject.Studio/wiki/BlackMagic-Support with the latest linux driver (14.4.2) and the source "Blackmagic Devices" is still not appearing.

Is the only workaround using the obs-studio package from apt or do that version have the same issue?

**Comment by Da-Boom** [comment]
That login wall is actually only superficial - on the bottom left of the login wall theres a "skip" button that can be used to skip that wall (which is good as my decklink device is an old one that they no longer list)

untill this gets fixed I'm gonna use the tytan652 release on the AUR, but I would love to switch permanently to flatpak in case I decide to distro hop to something other than Arch (it would be a lot easier to move my setup)

 @trwnh  
- i recommend the tytan652 release if you're on the AUR - it has a number of tweaks and comes with all the required plugins built  including browser source.

**Comment by cjuniorfox** [comment]
I followed the steps as detailed here: https://github.com/flathub/com.obsproject.Studio/wiki/BlackMagic-Support but I also had no luck.
Furthermore, I changed the builder script to the last version and published it on my GitHub repository: https://github.com/cjuniorfox/com.obsproject.Studio but still had no luck.
Executing at the terminal: `flatpak run com.obsproject.Studio` do I have those modules loaded:

```
info: ---------------------------------
info:   Loaded Modules:
info:     text-freetype2.so
info:     rtmp-services.so
info:     obs-x264.so
info:     obs-websocket.so
info:     obs-vst.so
info:     obs-transitions.so
info:     obs-outputs.so
info:     obs-filters.so
info:     obs-ffmpeg.so
info:     obs-browser.so
info:     linux-v4l2.so
info:     linux-pulseaudio.so
info:     linux-pipewire.so
info:     linux-jack.so
info:     linux-capture.so
info:     image-source.so
info:     frontend-tools.so
info:     decklink-output-ui.so
info:     decklink-captions.so
info: ---------------------------------
```
As can see, the modules:  `decklink-output-ui.so decklink-captions.so` was loaded, but I don't know if still have another module needed to be loaded.

**Comment by PhenaOfMari** [comment]
The past few days I've been stubbornly trying to find a way to get this working in one way or another (on Arch, using the AUR Decklink drivers). I've noticed that the flatpak build of obs still contains the decklink modules, but decklink.so fails to load because it can't find the libDeckLinkAPI.so (of course, sandbox and all).
```
libDeckLinkAPI.so: cannot open shared object file: No such file or directory
warning: A DeckLink iterator could not be created.  The DeckLink drivers may not be installed
warning: Failed to initialize module 'decklink.so'
```
The other thing I have noticed while inspecting the runtime using `flatpak enter com.obsproject.Studio ls [etc]` is that the libraries do exist in there at `/run/host/user/lib/libDeckLinkAPI.so`. Is there no way for the flatpak to detect that these are here and use them to allow the modules to properly load? I imagine its a similar story using flatpak obs with the honest to goodness blackmagic decklink .deb file on a debian based distro. If that is a possibility, maybe something like that can be a stopgap until proper flatpak extension support exists?

As a side note, I also sent a mail to BlackMagic support about a week ago asking about the possibility of them adding their libraries as a flatpak extension, to which they answered there were no current plans but it would be passed onto the development team for consideration. Maybe if there is enough of us voicing a want for this it could push them over the inertia of getting that set up. Probably not, but I can hope.

**Comment by tytan652** [comment]
> I've noticed that the flatpak build of obs still contains the decklink modules, but decklink.so fails to load because it can't find the libDeckLinkAPI.so (of course, sandbox and all).

I guess I let it enabled with the hope that BlackMagic add Decklink as a Freedesktop Platform Flatpak extension (`org.freedesktop.Platform.`).

> The other thing I have noticed while inspecting the runtime using `flatpak enter com.obsproject.Studio ls [etc]` is that the libraries do exist in there at `/run/host/user/lib/libDeckLinkAPI.so`. Is there no way for the flatpak to detect that these are here and use them to allow the modules to properly load? I imagine its a similar story using flatpak obs with the honest to goodness blackmagic decklink .deb file on a debian based distro. If that is a possibility, maybe something like that can be a stopgap until proper flatpak extension support exists?

TLDR: No

Flatpak must only rely on Flatpak runtimes and extensions.
Using host library is going against the concept of the Flatpak.
Also punching more hole in the sandbox must be avoided.

**Comment by PhenaOfMari** [comment]
> I guess I let it enabled with the hope that BlackMagic add Decklink as a Freedesktop Platform Flatpak extension (org.freedesktop.Platform.).

Oh well, here's hoping then. Having someone with a clear understanding of the creation effort and maintenance cost required in undertaking the extension path discuss that with the BlackMagic folks is probably the only way its going to sway them into doing it.

It was hard enough finding an good internal capture card with any linux support, it would be nice if I didn't have to do that again. Thanks for the info, anyway.
